### DISCUSSION TITLE
Context menu / right-click menu best practices?

### DISCUSSION THREAD

**Discussion by LoganDark** [tailwindlabs/headlessui#649] [Help]
Hello, I'm just getting started with React and Headless UI, and I'm not sure which Headless UI component would be best for a right-click menu. Perhaps Menu or Popover? Are there any good libraries to help with this as well?

**Comment by LoganDark** [comment]
Looks like there's no way to access the internal state of the Menu at least. No way to open it to activate the FocusTrap and no way to have it close when you click outside. So I will have to implement both myself...

It's a real shame that nobody can seem to figure out how to use this library to achieve a simple context menu. I wish the building blocks of headlessui were separate packages so I could make my own components with keybinds and focustraps and whatnot, but those are all completely private implementation details because _of course they are_.

**Reply by jordn** [reply]
+1

**Reply by filonik** [reply]
I am having similar issues figuring out best practices on how to use Popover. (See: https://github.com/tailwindlabs/headlessui/discussions/992)

It is a bit discouraging that there is not a lot of information on how to use these components. The existing examples often just show very basic use cases, and if those aren't exactly what you need, then you are out of luck. I understand that documentation doesn't write itself, but unfortunately there does not seem to be a lot of active discussion around here either.

I really love the idea of Headless UI in principle, but in practice I do find it a bit limited. I hope that it will grow, but at the moment it is very tempting to switch to some of the more fully featured component libraries.

**Reply by LoganDark** [reply]
I don't think the problem is that there's a lack of documentation regarding these components, but rather that the documented use cases are the **only** use cases that are supported, or even _possible_. For example, if it's shown how to use a button to open a popover, that is the **only** way to open the popover. If it's shown that you can click on tabs to change which one is open, that's the **only** way to switch tabs. The state is completely encapsulated - there is no option to pass it down from above; only to consume it (usually).

There's no flexibility here. There's no customizability. I find myself reimplementing basic things all the time, going down to lower-level primitives from other libraries to fill in the gaps that Headless UI doesn't fill. Sorry, I can't use the Menu component for my context menu - it's opened by a right click, which isn't a left click on a button. I have to pull in a focus trap from another library, because issues like #966 do not get fixed, and reimplement arrow key navigation (among others) myself.

If you open an issue regarding using the library for a use case that isn't supported, you get no help. You get told that the library is not designed for that, that that component shouldn't be used that way, etc. That that is the way that Headless UI should do things and is going to continue to do things, because in the end, this is a library for hobby interfaces and demos. Not real applications.

That, I feel, is the fatal flaw with this library that makes it a pain to use. The fact that the functionality is so limited, that it's so opinionated about the correct way to do things (without providing correct ways to do a lot of things), and that bugs don't get fixed when they aren't applicable to a supported use case.

I don't particularly like the way this library is run, and I've had bad experiences in the past with other tailwindlabs projects as well. I love Tailwind, and I like using Headless UI when it _actually_ suits my use case, which is why I'm prepared to put up with this, but I feel like it's a bad sign that this isn't sustainable. I think it's time to reconsider the direction that Headless UI is going, whether it should either become a _better set of building blocks_, or take the time to cover use cases like context menus, autocompletes and so on _The Correct Way_.

I use Tailwind and Headless UI in a _real application_ that is actually going to be sold to users - we have needs that this library _does not fill_ and _does not attempt to support in any way_. One or the other would be fine, but both? Come on.

**Reply by LoganDark** [reply]
In regards to other component libraries, Headless UI is almost the only one of its kind. It's the only one that does not bring its own classes and its own CSS. You can absolutely find other libraries for individual components - for example, I plan on adding downshift in the future - but not a toolkit (no matter how small). Especially since things like radio buttons don't usually come in their own libraries. 

It fits a different niche than full-featured libraries like ant/material design, which makes it even more sad that it does a subpar job of filling said niche. There aren't many alternatives out there (certainly none that I can list off the top of my head).

**Reply by AdnanTheExcellent** [reply]
+1 would love to see the ability to use a headlessui component as a right clickable context menu. There seems to be a pretty big gap in good and maintained context menu libraries for javascript. Most of the ones i find on github are outdated and / or unmaintained.

**Reply by LoganDark** [reply]
It doesn't look like headlessui is looking into supporting more use cases. Just their own. Same with heroicons - they're not even accepting pull requests.

Tailwind seems to be the only project that is being actively developed. That, and Tailwind UI, their paid offering.

**Reply by inwardmovement** [reply]
@LoganDark if you're looking for flexibility/customizability, you should check out [AlpineUI](https://alpinejs.dev/components).

**Reply by LoganDark** [reply]
> @LoganDark if you're looking for flexibility/customizability, you should check out [AlpineUI](https://alpinejs.dev/components).

I left the team I was working with about 2-3 months ago.

Even if I were still on the project, AlpineUI wouldn't be an option for many reasons. One of those reasons is that it's not React-compatible, so I'd have to create manual wrappers around every component. It's not a huge deal in terms of programmer effort, but the fact that it does so much DOM manipulation outside of React is pretty much a no-no. Libraries like [FlexLayout](https://github.com/caplin/FlexLayout) try hard to treat React components correctly even though they are mostly vanilla JS, but AlpineUI does not, and does not properly support the props system.

On top of that, the price isn't something I (or the team) would've been willing to pay. We developed the app on a budget of $0, and it eventually got scrapped anyway.

With that said, HeadlessUI is not going to be an option either in the future. It includes nice primitives like focus utilities, seem to be higher quality than any other library out there - too bad they're private implementation details, used to implement opinionated components that can't be configured to do what you want.

I think I'm done with React for the time being.

**Comment by designbyadrian** [comment]
Headless UI mention positioning their Popover component using Popper JS, see https://headlessui.com/react/popover#positioning-the-panel

In turn, Popper has "[virtual elements](https://popper.js.org/docs/v2/virtual-elements/)" which can be used as popper targets. Combine the knowledge of these two to achieve your goal.

**Reply by LoganDark** [reply]
I used Popper to position my context menu. I couldn't use Popover, though, because it can only be activated by left click on a button element. There is a huge encapsulation issue with Headless UI. It is *too* encapsulated.

**Reply by methyl** [reply]
@LoganDark check this out: https://github.com/tailwindlabs/headlessui/pull/1966. Feel free to use the fork, I'm going to maintain it for the time being (`"@headlessui/react": "BusinessModelFoundry/headlessui"` should work)

**Reply by LoganDark** [reply]
@methyl Unfortunately I no longer work on the piece of software that I needed Headless UI for, but it's great to see that you're taking steps towards making it more suitable for these use-cases in the future!

**Comment by x-ji** [comment]
You can just put a React ref on the `Menu.Button` (which is empty and thus doesn't show up in the UI) and trigger the click event via the ref upon the `onContextMenu` event (i.e. upon right click) of the target element. I don't see it as a huge "encapsulation issue" as you've been claiming here.

https://react.dev/learn/manipulating-the-dom-with-refs

**Reply by LoganDark** [reply]
What button?

Am I just supposed to create a hidden button just so I can call `.click()` on it when a user opens the context menu?

Respectfully, *what the fuck?*

**Reply by x-ji** [reply]
lmao why the hell are you so aggressive towards somebody posting a solution to the topic under discussion? This is completely uncalled for.

**Reply by LoganDark** [reply]
> lmao why the hell are you so aggressive towards somebody posting a solution to the topic under discussion? This is completely uncalled for.

That wasn't meant to be aggressive, I just don't understand why I should have to create and click a hidden button to open a context menu.

Anyway, I'm unsubscribing from this issue since I no longer work on that project and Tailwind Labs doesn't seem interested in addressing this.

**Reply by sho052** [reply]
@LoganDark And I don't think there's a need for you to block me from replying here either, as I was still typing out a longer response on the topic and didn't engage in personal attack or whatnot. (This is another account I have. Let me finish up what I had been typing here.)

I mean, I can understand it if you don't want to have a hidden button as it doesn’t conform to the DOM semantics. Also, to you this solution was probably obvious, but not what you needed at all, so I can imagine if you'd be somewhat annoyed at the way I explained it. If so, my bad. I hope you can understand though, this might be obvious to you, but not necessarily to others with less experience who had the same question while using HeadlessUI, like me who came across this post while searching online.

While I understand if you don't want to use products from Tailwind Labs anymore as you explained here (it's your freedom to choose and I also can't fix this issue), somebody else might still find this solution/workaround useful if they choose to stick with HeadlessUI for their project, regardless of how inelegant or "wrong" it is. This question still has value to the public even though you as the original submitter is no longer interested in it. Respectfully, I don't think you need to say "Respectfully, what the fuck?" here and react in such a way. You could have made your point normally and I would have taken it just fine.

**Reply by LoganDark** [reply]
@sho052 please leave me alone

**Comment by junwen-k** [comment]
I know this is not related to Headless UI, but do check out Radix UI's [Context Menu](https://www.radix-ui.com/primitives/docs/components/context-menu) as an alternative, it is fully accessible.

I switched from Headless UI to Radix UI ever since, because Radix UI have much more components to work with. :)
